daemon/cmd/ipam.go

```diff
@@ -308,58 +308,89 @@ func (d *Daemon) allocateDatapathIPs(family types.NodeAddressingFamily, fromK8s,
+		// Coalescing multiple CIDRs. GH #18868
+		if option.Config.EnableIPv4Masquerade &&
+			option.Config.IPAM == ipamOption.IPAMENI &&
+			result != nil &&
+			len(result.CIDRs) > 0 {
+			result.CIDRs = coalesceCIDRs(result.CIDRs)
+		}
 
-			// In ENI and AlibabaCloud ENI mode, we require the gateway, CIDRs, and the ENI MAC addr
-			// in order to set up rules and routes on the local node to direct
-			// endpoint traffic out of the ENIs.
-			if option.Config.IPAM == ipamOption.IPAMENI || option.Config.IPAM == ipamOption.IPAMAlibabaCloud {
-				var err error
-				if d.healthEndpointRouting, err = parseRoutingInfo(result); err != nil {
-					log.WithError(err).Warn("Unable to allocate health information for ENI")
-				}
+		if result.IP != nil {
+			node.SetEndpointHealthIPv4(result.IP)
+		}
+		log.Debugf("IPv4 health endpoint address: %s", result.IP)
+
+		// In ENI and AlibabaCloud ENI mode, we require the gateway, CIDRs, and the ENI MAC addr
+		// in order to set up rules and routes on the local node to direct
+		// endpoint traffic out of the ENIs.
+		if option.Config.IPAM == ipamOption.IPAMENI || option.Config.IPAM == ipamOption.IPAMAlibabaCloud {
+			if d.healthEndpointRouting, err = parseRoutingInfo(result); err != nil {
+				log.WithError(err).Warn("Unable to allocate health information for ENI")
 			}
 		}
+	}
 
-		if option.Config.EnableIPv6 {
-			result, err := d.ipam.AllocateNextFamilyWithoutSyncUpstream(ipam.IPv6, "health", ipam.PoolDefault())
+	if option.Config.EnableIPv6 {
+		var result *ipam.AllocationResult
+		var err error
+		healthIPv6 = node.GetEndpointHealthIPv6()
+		if healthIPv6 != nil {
+			result, err = d.ipam.AllocateIPWithoutSyncUpstream(healthIPv6, "health", ipam.PoolDefault())
 			if err != nil {
-				if healthIPv4 := node.GetEndpointHealthIPv4(); healthIPv4 != nil {
+				log.WithError(err).WithField(logfields.IPv6, healthIPv6).
+					Warn("unable to re-allocate health IPv6, a new health IPv6 will be allocated")
+				healthIPv6 = nil
+			}
+		}
+		if healthIPv6 == nil {
+			result, err = d.ipam.AllocateNextFamilyWithoutSyncUpstream(ipam.IPv6, "health", ipam.PoolDefault())
+			if err != nil {
+				if healthIPv4 != nil {
 					d.ipam.ReleaseIP(healthIPv4, ipam.PoolDefault())
 					node.SetEndpointHealthIPv4(nil)
 				}
-				return fmt.Errorf("unable to allocate health IPs: %s, see https://cilium.link/ipam-range-full", err)
+				return fmt.Errorf("unable to allocate health IPv6: %s, see https://cilium.link/ipam-range-full", err)
 			}
+		}
 
-			// Coalescing multiple CIDRs. GH #18868
-			if option.Config.EnableIPv6Masquerade &&
-				option.Config.IPAM == ipamOption.IPAMENI &&
-				result != nil &&
-				len(result.CIDRs) > 0 {
-				result.CIDRs = coalesceCIDRs(result.CIDRs)
-			}
+		// Coalescing multiple CIDRs. GH #18868
+		if option.Config.EnableIPv6Masquerade &&
+			option.Config.IPAM == ipamOption.IPAMENI &&
+			result != nil &&
+			len(result.CIDRs) > 0 {
+			result.CIDRs = coalesceCIDRs(result.CIDRs)
+		}
```

Comment:
These operations don't seem to have any effect, as `result.CIDRs` is then never used. This was already the case before your PR, but it may still be worth a cleanup while being here. My undestanding is that this is not required at the moment given that [IPv6 is currently not supported in ENI mode](https://github.com/cilium/cilium/issues/18405).

Reply:
Done, these unused operations has been cleaned up.